### DISCUSSION TITLE
Subtracting a value from itself equals not zero

### DISCUSSION THREAD

**Discussion by flawrist** [PaddiM8/kalk#132] [Q&A]
Kalker 2.0.4 on Windows 7 x64.

``` console
$ qalc 128,36+126,86-255,22
128,36 + 126,86 - 255,22 = 0       <--- Well done, as expected.

$ kalker 128.36+126.86-255.22
1.4210854715×10^-14 ≈ 0            <--- Whoa! I wasn’t expecting that.
```

**Comment by PaddiM8** [comment]
Some more to try:

`>> Σ(n=0, 720, 1÷n!)`  
`>> 4.4142135624`
`>> 3/4 + 5/7`

**Reply by PaddiM8** [reply]
Unfortunately there will always be some rounding errors so it adds some estimations like that as a clarification. In your case, it could be either 0 or almost 0 so it shows both haha

**Reply by flawrist** [reply]
Well, `kalker 1.5+1.5-3` returns `0` as expected, so I don't understand why my first example returns some scary sequence instead. This may have something to do with the way the machine calculates things, but the result clearly does not reflect what we were taught in school ([x-x=0](https://philosophy.stackexchange.com/a/82982)).

**Reply by PaddiM8** [reply]
Well it's a bit unpredictable, but it happens in some cases due to the nature of floating point numbers. In this specific case, it could technically be avoided, but the libraries for the Rust programming language are lacking unfortunately. I'm hoping to do it in the future though! But for now, it simply means that there were some tiny rounding errors and that the result is either that or 0.

**Reply by flawrist** [reply]
Dear @graydon, how come the language you designed doesn't offer a workaround for the [floating point problem](https://stackoverflow.com/questions/2002821/why-does-this-subtraction-not-equal-zero)? Look what happened here. Rust-based calculator outputs a puzzling result that raises anxiety levels and undermines confidence in the technology. Now I'm fiddling with the Citizen pocket calculator, which has served me for decades, to calm down a little.

**Reply by graydon** [reply]
Rust is a systems language and its default types are hardware-supported machine types. I did want to have (and originally had) decimal types in the standard library and as first class integrated types but I was overruled by others.

That said, there are plenty of general decimal libraries in Rust and if you want decimal arithmetic you can absolutely get it. You might like https://crates.io/crates/rust_decimal or https://crates.io/crates/decimal

**Reply by PaddiM8** [reply]
Yep it does exist. I am just waiting for one that has the features (arbitrary precision, wasm support, pure-rust, a decent amount of mathematical functions). There was some new one but it wasn't quite there yet last I checked.
